addons/payment_paypal/controllers/main.py

```diff
@@ -46,20 +47,33 @@ def paypal_return_from_checkout(self, **pdt_data):
         retrieved and with it the transaction which will be immediately post-processed.
         """
         _logger.info("handling redirection from PayPal with data:\n%s", pprint.pformat(pdt_data))
-        if not pdt_data:  # The customer has canceled or paid then clicked on "Return to Merchant"
-            pass  # Redirect them to the status page to browse the (currently) draft transaction
-        else:
             # Check the origin of the notification
-            tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
-                'paypal', pdt_data
-            )
-            try:
-                notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
-            except Forbidden:
-                _logger.exception("could not verify the origin of the PDT; discarding it")
-            else:
+        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
+            'paypal', pdt_data
+        )
+        try:
+            notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
+        except Forbidden:
+            _logger.exception("could not verify the origin of the PDT; discarding it")
+        else:
                 # Handle the notification data
-                tx_sudo._handle_notification_data('paypal', notification_data)
+            tx_sudo._handle_notification_data('paypal', notification_data)
+
+        return request.redirect('/payment/status')
+
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET', 'POST'], csrf=False,
+        save_session=False
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancellation PayPal with data:\n%s", pprint.pformat(pdt_data))
+        tx_sudo = request.env['payment.transaction'].sudo().search([
+            ('reference', '=', transaction_reference),
+            ('provider_code', '=', 'paypal'),
+        ])
```

Comment:
Could you elaborate a bit on this? Because in normal case we check if the reference we got from paypal exists in odoo, and in canceled case the reference is what we give so it always will be in there right?

Reply:
You're right, and your code works and is good, but `_get_tx_from_notification_data` was created purposefully for that. This is a method from the payment engine that is overridden in every payment provider to find the transaction based on the data sent back by the provider. 

If you look closely at `_handle_notification_data`, you'll see that the first call is also to `_get_tx_from_notification_data`. Using `_get_tx_from_notification_data` seems then the right choice to ensure that the search domain will be the same. 

Basically, I invite you to always use existing methods to keep the responsibility centralized and avoid duplicating code.